### PR TITLE
Redesign specs

### LINKED ISSUE #4: Push Redesign Docs to the docs folder
I think someone should push all the redesign docs to the docs folder of this repo or at least create a PR.

### PR DESCRIPTION
**Pull request by xuhang57**
This fixes #4

This is a previous effort by CCI-MOC/ims#154 @chemistry-sourabh

I haven't copied over the rest API documentation since we are about to rewrite it.

fwiw, M2-Architecture-Idea.png is using the BMI to refer to M2, which could confuse other contributors who are not familiar with BMI/IMS. Could be an easy fix for the person who has the original source file of that picture.

### CODE REVIEW

**Comment by chemistry-sourabh** [comment]
@xuhang57 I think it is better to put it in docs/design-specs or some other folder.

Also push the rest-api doc from the redesign PR as it is part of the design-specs. 

We will probably write a different rest-api.md that will go in doc/

**Comment by djfinn14** [comment]
Are we going to change the "BMI Core" to "M2 Core" like @xuhang57  mentioned? If not, I will approve this

**Comment by xuhang57** [comment]
yeah, we should at least clarify that. Should we use BMI or M2? Or we use both?

**Comment by xuhang57** [comment]
@chemistry-sourabh do we need this: m2-db.mwb ?

**Comment by chemistry-sourabh** [comment]
That is the MYSQL Workbench file for the schema, so yes.

**Comment by chemistry-sourabh** [comment]
rest_api.md is not updated for new schema so don't merge.

**Comment by xuhang57** [comment]
@chemistry-sourabh  I copied it from: https://github.com/chemistry-sourabh/ims/blob/48fc749ab94bc2a86ddc8c318c66c12f0d01ba43/docs/redesign-specs/REST/Api.md

How should I update it?

File: docs/design-specs/rest_api.md [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+
+***
+### update-tag
+
+Update attributes of a tag.
+
+**Supported attributes**
+* name
+
+`PATCH /provisionedInstance/<provisionedInstanceId>/tag/<tagId>`
```
Comment (chemistry-sourabh) [inline docs/design-specs/rest_api.md]: What I just realized is that this type of calls for tag doesn't make sense.

tagId is going to be unique and independent from provisionInstanceId. So It doesn't make sense to put it under provisionedInstance.

It should be /tag/<tagId>

Or the other option is tagId should start from 1 for every unique provisionedInstanceId

Comment (xuhang57) [reply docs/design-specs/rest_api.md]: @chemistry-sourabh so as discussed, we are still going to keep it, correct?

**Comment by xuhang57** [comment]
@chemistry-sourabh is there a way I can find the most updated Schema so I can update the rest_api.md

**Comment by nasibehteimouri** [comment]
@xuhang57 I think I do not hav access to upload the updated schema. Can you please find the attached (schema + photo of core tables). Thanks. 

[m3-db.zip](https://github.com/CCI-MOC/M2/files/1734017/m3-db.zip)
core

File: docs/design-specs/database_description.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
+## Core
+Core contains the tables (quota, image, provisioned instance, and tag) that are manipulated by user. 
+
+#### Quota
+This table contains information about the user organization or project (that we called it entity).
+* id: primary key.
+* entityId: user organization/project identifier. 
+* quota: storage allocated for user organization/project.
```
Comment (nasibehteimouri) [inline docs/design-specs/database_description.md]: * current_usage: the current amount of space used by the entity.

File: docs/design-specs/database_description.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#### Image
+This table contains information about the golden image(s) that an entity uses to provision node instance(s).
+* id: primary key.
+* name: entity's specified name for the image. 
+* entityId: id of the entity to which the image belong. 
+* type: type of the image (e.g. QCOW).
+* isPublic: visibility of the image that could be public or private. 
+* isSnapshot: if the image is snapshot or not. 
+* creationTime: which time the image is created?
+
```
Comment (nasibehteimouri) [inline docs/design-specs/database_description.md]: provisionedInstanceId: id of the provisionedInstance which is provisioned via this image.

File: docs/design-specs/database_description.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+* id: primary key.
+* entityId: id of the entity to which the node belong. 
+* creationTime: which time the node instance is provisioned?
+* type: type of provisioning that could be virtual or bare metal. 
+* name: entity's specified name for the provisioned instance.
+* networkdId: VLAN on which the node is provisioned. 
+
+#### Tag
+This table contains information about the checkpoints of the entity's provisioned node instance(s).
+* id: primary key.
```
Comment (nasibehteimouri) [inline docs/design-specs/database_description.md]: id: primary key which is globally unique.

File: docs/design-specs/rest_api.md (hunk 1 above)
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `PATCH /tag/<tagId>`

File: docs/design-specs/rest_api.md [hunk 5]
```diff
@@ -0,0 +1,750 @@
...
+
+**Request Body Example**:
+
+    {
+        "name": "tag-01"
+    }
+
+**Parameters**:
+* `provisionedInstanceId`: Id of the instance
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: it must be 
'tagId': Id of the tag

File: docs/design-specs/rest_api.md [hunk 6]
```diff
@@ -0,0 +1,750 @@
...
+**Authorization**: User / Admin
+
+**Response (on success)**: No Body
+
+***
+### delete-tag
+
+Delete an existing tag.
+
+`DELETE  /provisionedInstance/<provisionedInstanceId>/tag/<tagId>`
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `DELETE  /tag/<tagId>`

File: docs/design-specs/rest_api.md [hunk 7]
```diff
@@ -0,0 +1,750 @@
...
+
+**Request Body**: No Body
+
+#### Parameters:
+* `provisionedInstanceId`: Id of the instance
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: provisionedInstanceId must be removed, as tag is globally unique.

File: docs/design-specs/rest_api.md [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+**Authorization**: User/Admin
+
+**Response (on success)**: No Body
+
+***
+### list-tags
+
+List all tags for a instance.
+
+`GET /provisionedInstance/<provisionedInstanceId>/tag`
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `GET /tag/search?provisonedInstanceId=xxx`

File: docs/design-specs/rest_api.md [hunk 9]
```diff
@@ -0,0 +1,750 @@
...
+
+**Response Parameters**:
+* `tags`: List of all tags created by the user
+
+***
+### show-tag
+
+Show details of a tag.
+
+`GET /provisionedInstance/<provisionedInstanceId>/tag/<tagId>`
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `GET /tag/<tagId>`

File: docs/design-specs/rest_api.md [hunk 10]
```diff
@@ -0,0 +1,750 @@
...
+
+**Request Body**: No Body
+
+**Parameters**:
+* `provisionedInstanceId`: Id of the instance
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: no need for provisonedInstanceId.

File: docs/design-specs/rest_api.md [hunk 11]
```diff
@@ -0,0 +1,750 @@
...
+
+**Response Parameters**:
+* `name`: Name of the tag
+
+***
+### flatten-tag
+
+Create deep copy from tag.
+
+`POST /provisionedInstance/<provisionedInstanceId>/tag/<tagId>/flatten`
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `POST /tag/<tagId>/flatten`

File: docs/design-specs/rest_api.md [hunk 12]
```diff
@@ -0,0 +1,750 @@
...
+
+**Request Body Example**:
+
+    {
+	  "name": "flattened-tag-01"
+    }
+
+**Parameters**:
+* `provisionedInstanceId`: Id of the instance
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: provisonedInstanceId must be removed.

File: docs/design-specs/rest_api.md [hunk 13]
```diff
@@ -0,0 +1,750 @@
...
+    }
+
+**Parameters**:
+* `entityId`: Id of the entity whoes quota needs to be updated
+* `quota`: The new quota of the user (In GBs)
+
+**Authorization**: Admin
+
+**Response (on success)**: No Body
+
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: I think there is a need for an API to read current_usage. not sure about it, as "show" (in bellow) must be enough.

File: docs/design-specs/rest_api.md [hunk 14]
```diff
@@ -0,0 +1,750 @@
...
+**Parameters**:
+* `entityId`: Id of the entity
+
+**Authorization**: Admin
+
+**Response (on success)**:
+
+    {
+                "entityId": 2112,
+                "quota": 10.0
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: current_usage mus be added.

"current_usage": 5.5

File: docs/design-specs/rest_api.md [hunk 15]
```diff
@@ -0,0 +1,750 @@
...
+    }
+
+**Response Parameters**:
+* `entityId`: Id of the entity
+* `quota`: Datastore quota allocated to entity in GBs
+
```
Comment (nasibehteimouri) [inline docs/design-specs/rest_api.md]: `current_usage`: current space used by the entity.

**Comment by chemistry-sourabh** [comment]
@nasibehteimouri here are the changes in schema

1. Stick to CamelCase Everywhere, like current_usage write it as currentUsage.
2. Quota table, I think you can make the entityId the primary key.
3. Remove the additional entity_id in provisionedInstance and Image tables.
4. Change entityId type to VARCHAR(100)
5. The foreign key from provisioned instance to image was not done correctly. A new column called parentImageId needs to be added to provisionedInstance table that is a foreign key.
6. isActive should be removed in Tag table.
7. You can remove the id_Unique index in Tag table as id is a primary key and it is unique by default.
8. Add Unique Index for name_entityId in ProvisionedInstance and Image.

Thats it for the schema

**Comment by chemistry-sourabh** [comment]
@nasibehteimouri you can probably send @xuhang57 the changed rest_api.md file on slack and he can update it.

**Comment by chemistry-sourabh** [comment]
Guys please address comments in line 42 and line 76 also, missed them in meeting.

**Comment by nasibehteimouri** [comment]
@chemistry-sourabh  Thanks for the comment. I'll send the updated files to Lucas.

**Comment by xuhang57** [comment]
@chemistry-sourabh @nasibehteimouri  I just updated the rest API docs. Please review.

**Comment by naved001** [comment]
Is there any reason to not merge this?

**Comment by xuhang57** [comment]
Guess we could.

**Comment by naved001** [comment]
merging this since this has 2 approvals.

**Pull request merged**